python/GafferSceneUI/RenderPassEditor.py

```diff
@@ -898,3 +930,270 @@ def __combiner( results ) :
+
+		if self.__currentRenderPass is not None :
+			self.__busyWidget.setVisible( False )
+			self.__updateMenuButton()
+
+	def _updateFromEditable( self ) :
+
+		self.__menuButton.setEnabled( self._editable() )
+
+	def __acquireGlobalsPlug( self ) :
+
+		scriptWindow = GafferUI.ScriptWindow.acquire( self.getPlug().node().scriptNode() )
+		return scriptWindow.getLayout().settings()["__renderPassMenuAdaptedScene"]["globals"]
+
+	def __setDisplayGrouped( self, grouped ) :
+
+		self.__displayGrouped = grouped
+
+	def __setHideDisabled( self, hide ) :
+
+		self.__hideDisabled = hide
+
+	def __menuDefinition( self ) :
+
+		result = IECore.MenuDefinition()
+
+		result.append( "/__RenderPassesDivider__", { "divider" : True, "label" : "Render Passes" } )
+
+		renderPasses = self.__renderPasses.get( "enabled", [] ) if self.__hideDisabled else self.__renderPasses.get( "all", [] )
+
+		if self.__renderPasses is None :
+			result.append( "/Refresh", { "command" : Gaffer.WeakMethod( self.__refreshMenu ) } )
+		elif len( renderPasses ) == 0 :
+			result.append( "/No Render Passes Available", { "active" : False } )
+		else :
+			groupingFn = GafferSceneUI.RenderPassEditor.pathGroupingFunction()
+			prefixes = IECore.PathMatcher()
+			if self.__displayGrouped :
+				for name in renderPasses :
+					prefixes.addPath( groupingFn( name ) )
+
+			for name in sorted( renderPasses ) :
+
+				prefix = "/"
+				if self.__displayGrouped :
+					if prefixes.match( name ) & IECore.PathMatcher.Result.ExactMatch :
+						prefix += name
+					else :
+						prefix = groupingFn( name )
+
+				result.append(
+					os.path.join( prefix, name ),
+					{
+						"command" : functools.partial( Gaffer.WeakMethod( self.__setCurrentRenderPass ), name ),
+						"icon" : self.__renderPassIcon( name, activeIndicator = True ),
+						"description" : self.__renderPassDescription( name )
+					}
+				)
+
+		result.append( "/__NoneDivider__", { "divider" : True } )
+
+		result.append(
+			"/None",
+			{
+				"command" : functools.partial( Gaffer.WeakMethod( self.__setCurrentRenderPass ), "" ),
+				"icon" : "activeRenderPass.png" if self.__currentRenderPass == "" else None,
+			}
+		)
+
+		result.append( "/__OptionsDivider__", { "divider" : True, "label" : "Options" } )
+
+		result.append(
+			"/Display Grouped",
+			{
+				"checkBox" : self.__displayGrouped,
+				"command" : lambda checked : self.__setDisplayGrouped( checked ),
+				"description" : "Toggle grouped display of render passes."
+			}
+		)
+
+		result.append(
+			"/Hide Disabled",
+			{
+				"checkBox" : self.__hideDisabled,
+				"command" : lambda checked : self.__setHideDisabled( checked ),
+				"description" : "Hide render passes disabled for rendering."
+			}
+		)
+
+		return result
+
+	def __refreshMenu( self ) :
+
+		self.__busyWidget.setVisible( True )
+
+	def __setCurrentRenderPass( self, renderPass, *unused ) :
+
+		for plug in self.getPlugs() :
+			plug.setValue( renderPass )
```

Comment:
This definitely feels like it should be undoable when we're editing in the Settings window. It does feel a bit less like it should be undoable when editing in the main menu though. Should we make it undoable everywhere, or fudge it?

Reply:
I suppose the other precedent here is that `ScriptNode.frame` plug, which gets edited by the Timeline without creating an undo entry. But that's somewhat hidden, whereas in this case we're clearly making a direct edit to a plug on the ScriptNode.